FIX: Ensure backwards compatibility with old ENV variables

eviltrout opened the pull request:
Fast Rejection "aka Backscatter" will not execute unless the new
ENV variable is present. The original policy will work with both.

mpalmer commented on boot:
```diff
@@ -29,13 +29,18 @@ done
 
+if [ -z "$DISCOURSE_BASE_URL" ] && [ -z "$DISCOURSE_MAIL_ENDPOINT" ] ; then
+  echo "FATAL ERROR: You need to define DISCOURSE_BASE_URL or DISCOURSE_MAIL_ENDPOINT" >&2
+  exit 1
```
These look an awful lot like spaces, not tabs...

mpalmer commented on discourse-smtp-fast-rejection.rb:
```diff
@@ -13,7 +13,7 @@ def logger
 
 def fatal(*args)
-  logger.crit *args
+  logger.crit(*args)
```
Seems unnecessary to change this, but style consistency, I guess... :woman_shrugging:

eviltrout replied:
I run a ruby linter locally on save of files and fix issues it pops up. In this case, without the parentheses it's ambiguous about whether you want to multiply or expand args.

mpalmer replied:
Ooooooh, yeah... didn't think of that.  Definitely want some parens there, then...

mpalmer commented on discourse-smtp-fast-rejection.rb:
```diff
@@ -25,7 +25,10 @@ def main
-    fatal "env var %s is required", kw unless real_env[kw]
+    unless real_env[kw]
+      logger.debug "env variable #{kw} missing, fast rejection disabled"
+      real_env['FAST_REJECTION_DISABLED'] = true
```
This smells to me like it's time to put this into a proper class and flag this "lolidunno" mode with an instance variable.  Mangling the parsed environment to signal this feels like kicking a puppy.  A *really* cute puppy.

mpalmer commented on receive-mail:
```diff
@@ -25,10 +25,16 @@ def main
+  if real_env['DISCOURSE_MAIL_ENDPOINT'].nil? &&
+      real_env['DISCOURSE_BASE_URL'].nil?
+
+		fatal "DISCOURSE_MAIL_ENDPOINT or DISCOURSE_BASE_URL env var missing"
```
`s/ or / and /`

eviltrout commented:
Okay, following the advice to split it into an actual class, I did a big refactor:

- Created a base class for both receivers with common helpers
- The disabled flag is now an instance variable on the FastRejection class
- Added specs for everything.

mpalmer commented:
OMG specs... :heart_eyes: 

I'll do another round of code review on monday.

eviltrout commented:
Just a polite reminder I'm still waiting for this review :)

mpalmer commented on discourse-smtp-fast-rejection.rb:
```diff
@@ -1,111 +1,10 @@
 #!/usr/bin/env ruby
```
Oh, I forgot to mention this previously... any chance you could rip the `.rb` off the end of this?  Implementation details shouldn't be in program names.

mpalmer commented:
Congratulations!  You have survived the gauntlet of a womble code review.  :grinning: It all looks good to me.

eviltrout commented on spec/spec_helper.rb:
```diff
@@ -0,0 +1,20 @@
+  end
+
+  config.shared_context_metadata_behavior = :apply_to_host_groups
+end
```
Ah, that's because rspec autogenerated that file with spaces so my editor settings didn't catch it. Thanks.

tgxworld replied:
No problem 👍 Spaces and tabs are super easy to spot on Github. 8 spaces for a tab 😟

eviltrout commented:
@mpalmer Okay I've renamed the file and fixed the spaces in the file @tgxworld pointed out.

Merge if it looks good and I'll tackle the next step on Monday!
